### PR TITLE
End new persistence id search without relying on refresh interval

### PR DESCRIPTION
**Pull request by chbatey**
To be ported to the master branch

Two improvements;
* Check if the deadline for a new persistence id search has elapsed when a query is exhausted, not on the next refresh interval
* Schedule an extra refresh based on the new persistence id deadline if it is smaller than the refresh interval

Refs https://github.com/akka/akka-persistence-cassandra/issues/570

### CODE REVIEW

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala [hunk 1]
```diff
@@ -354,6 +339,25 @@ import com.datastax.driver.core.utils.UUIDs
         }
       }
 
+      private def abortMissingSearch(missing: LookingForMissing): Unit = {
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala]: Extracted method, no changes to content

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala [hunk 2]
```diff
@@ -577,18 +581,31 @@ import com.datastax.driver.core.utils.UUIDs
...
-            updateStageState(_.copy(missingLookup = stageState.missingLookup.map(m => {
-              val newBucket = TimeBucket(m.maxOffset, bucketSize)
-              m.copy(bucket = newBucket, queryPrevious = !newBucket.within(m.previousOffset))
-            })))
-            // a current query doesn't have a poll schedule one
-            if (!isLiveQuery())
-              scheduleOnce(QueryPoll, settings.refreshInterval)
+            val timeLeft = missing.deadline.timeLeft
+            if (timeLeft <= Duration.Zero) {
+              abortMissingSearch(missing)
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala]: ^ part one of the fix

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala [hunk 3]
```diff
@@ -577,18 +581,31 @@ import com.datastax.driver.core.utils.UUIDs
...
+                timeLeft.pretty)
+              updateStageState(_.copy(missingLookup = stageState.missingLookup.map(m => {
+                val newBucket = TimeBucket(m.maxOffset, bucketSize)
+                m.copy(bucket = newBucket, queryPrevious = !newBucket.within(m.previousOffset))
+              })))
+
+              // the new persistence-id scan time is typically much smaller than the refresh interval
+              // so do not wait for the next refresh to look again
+              if (timeLeft < settings.refreshInterval) {
+                scheduleOnce(QueryPoll, timeLeft)
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala]: ^ part two

File: core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala [hunk 4]
```diff
@@ -789,6 +793,56 @@ class EventsByTagStrictBySeqNoEarlyFirstOffsetSpec
   }
 }
 
+class EventsByTagLongRefreshIntervalSpec
```
Comment (chbatey) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: This wasn't caught in the previous test as we set the refresh interval very low to speed up tests

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala (hunk 2 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala]: It depends what kind of missing search is happening.

If it is a new persistence id then it is set to the usually small `new-persistence-id-scan-timeout` from `handleFirstTimePersistenceId`

If it is an actual gap, e.g. we see 1-4, then say 6 then it is set to the, much larger as we will fail the stream, `gap-timeout` set in `handleExistingPersistenceId`

Comment (patriknw) [reply core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala]: Ok, but abortMissingSearch will fail the stream also for the first case (handleFirstTimePersistenceId), or did I misread that?

Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala]: `abortMissingSearch` only fails the stream if `failIfNotFound` is set to true, which is only set in the second case.

I'll re-name that field in `LookingForMissing` to `gapDetected` to make it clearer (I hope).

**Comment by chbatey** [comment]
The new test failed in one of the builds: https://travis-ci.org/akka/akka-persistence-cassandra/jobs/607032979?utm_medium=notification&utm_source=github_status so i'll investigate that before merging this

**Comment by chbatey** [comment]
This needs porting to master as well

**Pull request merged**